### ISSUE TITLE
[Feature Idea] H.A.G. (Holographic Assembly Glove)

### ISSUE THREAD

**Issue by HipHopHuman** [TeamPneumatic/pnc-repressurized#842]
This is probably waaaaaaaay out of scope, so I won't be surprised if you reject it, but I've got an idea for a new feature that will make early-game crafting more convenient.

## The Idea

A glove that offers the player a more accessible (than the vanilla Crafting Table) & portable standard 3x3 recipe crafting mechanic, complete with fancy visual effects that fit the Iron Man theme, notes regarding balance and tiers of progression through upgrades.

## The Rationale

I've been playing a PneumaticCraft-only survival world (with the only other mods being for performance, slight QoL and decoration), and one thing that has bothered me a lot in the early game is that of the manual crafting situation. Every time I need to craft something, I have to place down a crafting table (which I have to always carry with me). If I don't have a crafting table on me, I either need to run back to base or craft a crafting table (which sometimes involves cutting trees) and place it down, whichever is the most convenient.

This led me to test out an idea, where I created a Drone program to teleport to me and place a Crafting Table down when I click a button on the remote. It worked, but it's a bit clunky. Regardless, I am surprised by just how much use I am getting out of this particular Drone program - I use it at least once every 30 minutes.

In other mods, I'm spoiled by more convenient crafting. Botania has the Assembly/Manufactory Halos, Charm has the 3x3 crafting button when you're carrying a crafting table, Actually Additions has the Crafting Table on a Stick, EvilCraft has the Exalted Crafters (one of which links with the Ender Chest and functions as a sort of "Project Table"), Mekanism has Robit and so on. 

I could just go ahead and install a "Portable Crafting" mod, but the problem with those is they're either *too* convenient (requiring just having a crafting table in your inventory and/or a Sign/Stick to craft it with) and feel too out-of-place, thematically.

This leads me to believe that PneumaticCraft might benefit from a more convenient but also more *balanced* portable crafting mechanic, one that might serve well to remove a chunk of tedium from the early game without messing with progression. So, I thought about how such a mechanic might work, keeping with the "Iron-Man" theme, and the Holographic Assembly Glove, or "H.A.G." for short, is what I came up with.

## The Item

It's a glove. Despite being a glove, it is not a Curio/Bauble, nor is it a piece of Armor. Mechanically, it functions like a tool, in that the player needs to equip it in either the off-hand or the main hand in order to use it.  When equipped, it looks like the player is "wearing it" (kind of like the Air Cannon does, in that it covers some of the player's arm).

Like all other Pneumatic tools, it requires Air Pressure to function, and like some Pneumatic tools, it can make use of upgrades.

## How the Crafting Mechanic could work

If the H.A.G. is equipped in either hand and has sufficient air, then the block that the player is currently looking at will have a blue (or any other color, maybe the same color as the HUD) 3x3 grid of lines super-imposed over it's top face (it could use the same rendering technique as the Logistics Frames). Thematically, this grid is a projection, made up of holographic light. To keep with the "light" theme, perhaps it could have an emissive texture (like Spider/Enderman eyes have), so that it "glows" in the dark. Visually, it could look something like this in-world: 

![image](https://user-images.githubusercontent.com/70091260/124392272-db0b1680-dcf4-11eb-847f-b20d1ffe5b71.png)

The way the crafting mechanic works is; the player can right-click each individual cell of the grid with an item to "place" the item into that cell. The items will be rendered (the same way they are rendered on the Display Table).

If the arrangement of items on the grid is *not* a valid crafting recipe, then the grid will turn red.

However, if the arrangement of items on the grid *is* a valid crafting recipe, the grid will stay it's original color, and a transparent "ghost" item entity of the recipe's output will be rendered above the grid. If the recipe has multiple conflicting outputs, then this "ghost" item will periodically change, cycling between each potential output. 

Right-clicking a cell on the grid with an item that matches/already exists in that cell will fill *all* pre-configured cells with 1 more item, sourced from the player's inventory.

With all items for a valid crafting recipe on the grid, the grid won't do anything on it's own. It needs to be "activated" (by right-clicking with at least one empty hand). In the case of recipe output conflicts, the chosen crafted output is determined by which "ghost" entity was rendered above the grid at the time of the activation right-click.

Once activated, the items in the outermost cells of the grid will start slowly moving toward the center cell of the grid (a process that takes some time, keeping this a little more balanced than other portable crafting solutions) - once all the items have all reached the center cell, they are consumed and the crafting output pops out in item entity form.

In the case of accidentally misclicking an item into the wrong cell, or just wanting to cancel, shift-right clicking while the H.A.G. is equipped will remove the rendered grid that the player is currently looking at from the world, and all input items currently on the grid will pop out in item form.

The grid itself will stay behind and remain bound to the same relative coordinate as the player, so the player has easy access to it should they equip the H.A.G. again.

Each time the grid performs a craft, it consumes a marginal amount of air from the H.A.G.

## What upgrades it could take

This is what I think might be appropriate for the H.A.G. with regards to upgrades:

- Volume Upgrades - bog-standard, these just increase the amount of air the H.A.G. can store internally. Accepts up to 25 for a 10x increase.
- Speed Upgrades - these will speed up the amount of time it takes for the outermost input items to travel to the center cell, at the cost of an exponential increase in air usage. The base speed takes 5 seconds. 6 speed upgrades makes the process instant.
- Inventory Upgrades - these will increase the amount of items each individual cell can hold by 1 (with the base amount being 1, and the max being 64)
- Range Upgrades - these will increase the amount of grids the player may have active at a time (with the base amount being 1, the max being 16)
- Dispenser Upgrade - This upgrade will change the mechanics of the holographic crafting grid somewhat, making it more useful for mass-crafting and better management of crafting projects with a lot of micro-crafts. I envision this as being the functional equivalent of Botania's Manufactory Halo, but it requires a bit of setup and doesn't run indefinitely. This is how the mechanics are changed by the inclusion of a Dispenser Upgrade (the max dispenser upgrades allowed is 1):
  - This upgrade benefits directly from the H.A.G. also having Inventory Upgrades installed. 
  - If the grid is placed on top of an inventory, it will "suck" all items from that inventory into pre-configured slots.
  - Once the grid crafting is activated with a bare hand, it will continuously craft, until all items in the grid's cells have been exhausted (even if the H.A.G. is unequipped during this process)
  - The air usage of the H.A.G is calculated differently. Instead of calculating the air once a craft is done, it pre-calculates the air cost based on how many crafts will occur during it's lifetime.  It does this by introspecting how many items are in each cell (and how many items are in the container below, if necessary).
  - The Dispenser-Upgraded H.A.G doesn't run indefinitely, but it can be *made* to run indefinitely if the container below it is being fed by automation at a faster rate than the crafting progress takes to craft an item. This can be used to automate compression of stuff like redstone and coal into blocks on mining trips if paired with an Aerial Interface.

## The Recipe

I haven't given much thought to what the recipe for this could look like, as long as it fits the theme and isn't *too* early or *too* late into progression. Perhaps a multi-step craft, involving Leather, String, Pneumatic Cylinders, an Air Canister and the UV Lightbox. That way it requires at least some basic progression in the mod to make, while being pre-PCB, which I think puts it at *just* the right "level" of game progression.

Perhaps something like this:

- Craft 4 Leather + String into a "Glove" crafting component (pattern: [L][L][L][ ][S][ ][ ][L][ ])
- Craft Glove + Crafting Table + UV Lightbox + Pneumatic Cylinder into H.A.G. (pattern: [P][U][P][P][C][P][ ][G][ ])

## Other thoughts

- If this idea were to be implemented, it'd be neat if Drones had a function where they could equip the H.A.G, which projects the same crafting grid on the block below them when they execute a Crafting Widget, just to make the whole Drone autocrafting thing more visually appealing.
- If this idea were to be implemented *in it's entirety*, then it would introduce the fact that PneumaticCraft is now aware of Recipe Conflicts, and the Crafting Widget might need to be adjusted to allow for manually choosing from potential crafting outputs for the sake of consistency.

## Conclusion

So, that's my idea. Let me know what you think :D

**Comment by desht** [comment]
Hmm.  It's a neat idea, but here's the big problem: it's far less convenient than just putting down a vanilla crafting table and using JEI to insert everything you need with one click, and then one more click to extract the result.

So I'm struggling to see how people would actually choose to use this over something that's available right at the very start of the game.

**Comment by HipHopHuman** [comment]
I can't remember which mod added it, I think it was Draconic Evolution or ExtraUtils, but there was a mechanic where pressing "C" while looking at a block's top face would put a 3x3 grid on it that could be used as an immersive crafting interface - that's where I got the base of this idea from. It may sound less convenient, but trust me, it was super convenient. 

That aside, the convenience factor in this particular design is a progression goal - it starts out being less convenient, but when you have the tool fully upgraded, you could do bulk crafting of several different item types much faster than you could with JEI+Clicking. 

Think of these grids as PNC flavored Tinker's Crafting Stations, with inventory slots. The difference is you can have up to 16 of them active at a time, but only have to carry one item. 

With the dispenser upgrade, you can also have them continuously craft their  configured recipe until they've exhausted their entire stock. A temporary set-and-forget-until-it's-ready mechanic.

With 16 different recipe grids each crafting a dozen or more items in parallel you could achieve a lot more in much shorter time than you could with the regular crafting table GUI and shift-clicking.

If you do want a GUI though, nothing prevents you from modifying this idea and making it just open the crafting table GUI, rather than inputting items manually.
